File: 010_reverse_str/program.nim

#import std/[math, strutils]
#[
import std/math


proc reverse_str(str: string): string =
  var str2: string = ""
  
  for n in str :
    str2 = n & str2
  result = str2


var hello: string = "hello"
var world: string = "world"
var passed: string = "Assertion passed!"

echo hello, " -> ", reverse_str(hello)
echo world, " -> ", reverse_str(world)
echo passed, " -> ", reverse_str(passed)

echo "done!"
]#


#import std/[math, algorithm]
import std/[math, unicode]

var hello: string = "hello"
var world: string = "world"
var passed: string = "Assertion passed!"

echo hello, " -> ", reversed(hello)
echo world, " -> ", world.reversed
echo passed, " -> ", passed.reversed

echo "done!"

File: 010_reverse_str/program.c

#include <stdio.h>
#include <string.h>
#include <assert.h>

void reverse_str(char* str){
  int next = 0;
  int last = strlen(str) - 1;
  char tmp;
  
  while(next < last){
    tmp = str[next];
    str[next] = str[last];
    str[last] = tmp;
    next++;
    last--;
  }
}

int main(void) {
  char hello[] = "hello";
  char world[] = "world";
  char passed[] = "Assertion passed!";
  
  printf("%s -> ", hello);
  reverse_str(hello);
  printf("%s\n", hello);
  
  assert( ! strcmp(hello, "olleh") );
  //assert( strcmp(hello, "olleh") == 0 );
  
  printf("%s -> ", world);
  reverse_str(world);
  printf("%s\n", world);
  
  assert( ! strcmp(world, "dlrow") );
  
  printf("%s -> ", passed);
  reverse_str(passed);
  printf("%s\n", passed);
  
  assert( ! strcmp(passed, "!dessap noitressA") );
  
  printf("done!\n");
  
  return 0;
}


